src/software/util/is_template_base_of/is_template_base_of.hpp

```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include <type_traits>
+
+/**
+ * Type trait to check whether Derived inherits from any instantiation
+ * of the class template Base.
+ *
+ * Taken from: https://stackoverflow.com/a/73034562
+ *
+ * TODO (#3242): Change is_template_base_of to use C++20 concepts
+ *
+ * @example See is_template_base_of_test.cpp
+ *
+ * @tparam Base the class template
+ * @tparam Derived the type to check
+ */
+template <template <class...> class Base, class Derived, class = void>
+struct is_template_base_of : std::false_type
+{
+};
+
+template <template <class...> class Base, typename... Ts>
+void test(Base<Ts...>&);
```

Comment:
what is this `test` function for?

Reply:
It's used in this specialization of `is_template_base_of`

https://github.com/UBC-Thunderbots/Software/blob/eabe705239de8a2824ae693ed131bb3784f8c8c2/src/software/util/is_template_base_of/is_template_base_of.hpp#L26-L31

I'm not entirely sure how this works, but I think we're checking the validity of passing in Derived into a `test` function that only takes instantiations of the `Base` template.